addons/pos_pinelabs/static/src/app/payment_pinelabs.js

```diff
@@ -0,0 +1,242 @@
+import { _t } from "@web/core/l10n/translation";
+import { PaymentInterface } from "@point_of_sale/app/payment/payment_interface";
+import { AlertDialog } from "@web/core/confirmation_dialog/confirmation_dialog";
+import { serializeDateTime } from "@web/core/l10n/dates";
+
+const REQUEST_TIMEOUT = 10000;
+const CANCEL_REQUEST_TIME_LIMIT = 125000;
+const { DateTime } = luxon;
+
+export class PaymentPinelabs extends PaymentInterface {
+    setup() {
+        super.setup(...arguments);
+        this.paymentInitialTime = null;
+        this.pollingTimeout = null;
+        this.inactivityTimeout = null;
+        this.queued = false;
+        this.payment_stopped = false;
+    }
+
+    send_payment_request(cid) {
+        super.send_payment_request(cid);
+        return this._processPinelabs();
+    }
+
+    pendingPinelabsPaymentLine() {
+        return this.pos.getPendingPaymentLine("pinelabs");
+    }
+
+    send_payment_cancel(order, cid) {
+        super.send_payment_cancel(order, cid);
+        return this._pinelabsCancel();
+    }
+
+    _callPinelabs(data, action) {
+        return this.env.services.orm.silent
+            .call("pos.payment.method", action, [[this.payment_method_id.id], data])
+            .catch(this._handle_odoo_connection_failure.bind(this));
+    }
+
+    _handle_odoo_connection_failure(data = {}) {
+        // handle timeout
+        const line = this.pendingPinelabsPaymentLine();
+        if (line) {
+            line.set_payment_status("retry");
+        }
+        this._showError(
+            _t(
+                "Unable to connect to the Odoo server. Please check your internet connection and try again."
```

Comment:
we already have an error handler for connection errors 
https://github.com/odoo/odoo/blob/4822ae2d5805a18eea49e0611aff363d2e43d142/addons/point_of_sale/static/src/app/utils/error_handlers.js#L34-L49